crates/ruff_python_formatter/src/other/f_string.rs

```diff
@@ -24,26 +28,96 @@ impl<'a> FormatFString<'a> {
 impl Format<PyFormatContext<'_>> for FormatFString<'_> {
     fn fmt(&self, f: &mut PyFormatter) -> FormatResult<()> {
         let locator = f.context().locator();
+        let comments = f.context().comments().clone();
+
+        let string = StringPart::from_source(self.value.range(), &locator);
 
-        let result = StringNormalizer::from_context(f.context())
+        let normalizer = StringNormalizer::from_context(f.context())
             .with_quoting(self.quoting)
-            .with_preferred_quote_style(f.options().quote_style())
-            .normalize(
-                &StringPart::from_source(self.value.range(), &locator),
-                &locator,
-            )
-            .fmt(f);
-
-        // TODO(dhruvmanila): With PEP 701, comments can be inside f-strings.
-        // This is to mark all of those comments as formatted but we need to
-        // figure out how to handle them. Note that this needs to be done only
-        // after the f-string is formatted, so only for all the non-formatted
-        // comments.
-        let comments = f.context().comments();
-        self.value.elements.iter().for_each(|value| {
-            comments.mark_verbatim_node_comments_formatted(value.into());
-        });
-
-        result
+            .with_preferred_quote_style(f.options().quote_style());
+
+        // If f-string formatting is disabled (not in preview), then we will
+        // fall back to the previous behavior of normalizing the f-string.
+        if !is_f_string_formatting_enabled(f.context()) {
+            let result = normalizer.normalize(&string, &locator).fmt(f);
+            self.value.elements.iter().for_each(|value| {
+                comments.mark_verbatim_node_comments_formatted(value.into());
+            });
+            return result;
+        }
+
+        // TODO(dhruvmanila): This could probably be simplified for Python 3.12 specifically
+        // as same quotes can be re-used inside an f-string.
+        let quotes = normalizer.choose_quotes(&string, &locator);
```

Comment:
What I was thinking was that we could probably reduce the scope of the function to only use the f-string literal elements and not the expression elements in 3.12. The reason being that same quotes can be re-used in 3.12. But, I think the current logic is correct, I'll remove the comment. It might be more involved to check if the scope can be reduced.